### ISSUE TITLE
Merging branches re-executes commits

### ISSUE THREAD

**Issue by Bertie2011** [desktop/desktop#9815]
### Describe the feature or problem you’d like to solve

Previously, when a branch is merged into another branch with GitHub Desktop, the merged commits would remember when they were made. If the source branch modified code, that the destination branch didn't touch, the same piece of code would be modified in the destination branch without any problems once the branches were merged for a second time.

Since a couple of weeks, a second merge from source branch to destination branch is resulting in merge conflicts between the old and new versions of the source branch. It looks like Git or GitHub Desktop thinks the changes from the merge are applied by the destination branch. Merging into the default branch using the website works just fine, but merging 2 other branches using the software creates the mess of conflicts.

Maybe I can clarify it with a step-by-step reconstruction:
(Previously)
1. Create line `A` on source
2. Merge source into destination
3. Destination now has line `A`
4. Change line `A` to `B` on source
5. Merge source into destination
6. Destination now has line `B`

(Now)
1. Create line `A` on source
2. Merge source into destination
3. Destination now has line `A`
4. Change line `A` to `B` on source
5. Merge source into destination
6. Destination now has conflict:
`>>>>>> HEAD`
`A`
`======`
`B`
`>>>>> source`

Again, I assume this happens because for some reason git or GitHub Desktop thinks `A` has been committed on destination, even though destination didn't touch it.

### Proposed solution

For the commits to not be re-applied on merge. Those unnecessary conflicts are tiring.

**Comment by rafeca** [comment]
Hey @Bertie2011!

Thanks for filling this issue and your detailed reproduction steps. The behaviour that you describe is indeed incorrect, but I've tried locally with a new repository and I'm not able to reproduce it following the same steps that you mention (also I'm not aware of any recent change in Desktop that could have caused a change in behaviour here...).

I've recorded my screen while trying to reproduce the issue, can you take a look at it and let me know if I understood the repro steps correctly?

![demo](https://user-images.githubusercontent.com/408035/82201084-1cc5db80-9900-11ea-9604-090cfe82f76f.gif)

**Comment by Bertie2011** [comment]
That's very odd.... Yes that seems to be the right way to reproduce it. Does GitHub Desktop have hidden merge settings somewhere? Maybe using git bash itself?

Hmmm, does GitHub desktop use the repo's merge settings in settings > options, if only 1 is checked?
I'm aware that this is turning into a generic Git/GitHub question, but perhaps you could help me still.

**Comment by rafeca** [comment]
Since Desktop uses `git` internally, some of your local `git` settings can indeed affect how Desktop does things.

In this specific case, the only config param that I can think of that can affect this process would be [`merge.ff`](https://git-scm.com/docs/merge-config#Documentation/merge-config.txt-mergeff), which allows `git` to avoid doing merges when it's not necessary.

I have the default setting here (which is enabled), can you check if you have overriden this setting?

**Comment by Bertie2011** [comment]
I just tested it and indeed GitHub Desktop (or git internally I guess) uses the merge behavior specified in Settings > Options, if only 1 option is ticked.
The owner of the repo in question for some reason decided to only allow rebase merges, resulting in old commits to be added as brand-new. I don't know his reasoning, but I know that picking squash merging or merge commit options solve it.

Thanks for your time anyways :)

**Comment by Bertie2011** [comment]
Hmmm maybe this would be more of use for others if it would be moved to a GitHub repo. Maybe it could be transferred?

**Comment by Bertie2011** [comment]
I just talked to the repo owner and apparently I was right on only 1 option being picked, but that option was Squash merge (which I tested to work just fine).... Is there anywhere else that could specify rebase merges to be the default???

**Comment by Bertie2011** [comment]
#3422 has been active until April, which is roughly when I started seeing the conflicts. The git config doesn't specify pull.rebase though :/

It's also worthy to note that merges into the default branch using the GitHub Pull Request system works just fine. It's only caused by merging branches on GitHub Desktop

**Comment by rafeca** [comment]
@Bertie2011 can you paste your `~/.gitconfig` file (the parts that do not contain personal info)? maybe there's some other setting that's interfering...

Just to confirm, are you able to reproduce the issue following the exact same steps that I showed in the gif in https://github.com/desktop/desktop/issues/9815#issuecomment-630083542 ?

**Comment by Bertie2011** [comment]
I must be seeing things.... Just that example doesn't create conflicts, but if I add another line on "destination" before the second edit+merge (just after step 3.) I do get conflicts. This could be due to a line ending being added to the end of the already existing line in "destination".

I don't have concrete examples of where it went wrong inside files, but I do have an example of what happened recently "cross file":
On the destination branch, I edited a couple files, which were moved and renamed on the source branch. I thought git was able to resolve this, aka apply the changes to the relocated files and merge without problems, but maybe this scenario would always break and I just remembered it wrong. Anyways, I ended up with both my changed files and the relocated files. I had to find the changes in the original file and re-apply them in the relocated files, essentially manually merging.

Git config file:
[filter "lfs"]
clean = git-lfs clean -- %f
smudge = git-lfs smudge -- %f
process = git-lfs filter-process
required = true

**Comment by Bertie2011** [comment]
Well, we found one culprit. Someone was rebasing the default branch into their own branche every time they wanted to update. Since everything else is most likely usage error, I'll close this issue.

Good to know that **rebasing recreates old commits as new** and if you don't fully know what you're doing, it should be avoided to prevent unnecessary conflicts. (Obvious for many, but I wanted to put it here as conclusion for anyone finding this issue).
